### DISCUSSION TITLE
Does compaction affects search results?

### DISCUSSION THREAD

**Discussion by RakeshRaj97** [milvus-io/milvus#35536] [Q&A and General discussion]
I have a database of size 1 billion entries and I inserted data using bulk insert API. Each of my parquet file has 5 million rows and I set the maxSegment size to be 4096. Once fully inserted, I noticed the number of segments created to be 1000 totalling to 1 Billion entries. I created an index with a nlist of 1024. After couple of days I noticed the number of segments to reduce and I see a lot of segments of size 10 million. I am now wondering whether this increase in segment size has an effect on the search result given that the nlist is not changed from 1024. 

I also notice a very high usage in data nodes and I guess it's all due to the background compaction tasks which I never asked for.

Should I disable the compaction in dataCoord?

**Comment by yhmo** [comment]
Could you provide more info:
- dimension value?
- the collection schema?
- how do you pass the parquet files to bulkinsert API? One file per call, or input all files in one call?

Normally, no need to disable compaction. Each time you call bulkinsert API, a bulkinsert job is generated, it reads data from parquet files and assembles the data to segments according to shard number, partitions, and segment max size. That means the data in one parquet file can be assigned to different segments, the size of the segments might be different. Compaction merges small segments to large segment which is expected. If the dimension is 128. and maxSegment size is 4096MB, then it is possible to merge to a 4GB+ segment with 10 million rows.

Compaction might **slightly** affect search recall rate if your search with high topk value. The smaller nprobe value, the greater impact to recall rate.
In milvus, each segment has an independent index. 
Let's say we have 10 million vectors. The nlist is 1024.
- Before compaction, the 10 m vectors are stored in 10 segments, 1 million per segment. For each segment, the index has 1024 clusters which means 1000 vectors per cluster.
Now we search with nprobe=5, then 5 clusters are scanned, about 5000 vectors are computed per segment, totally 50000 vectors are computed for the 10 segments.

- After compaction, the 10 m vectors are stored in one segment. The index has 1024 clusters, 10000 vectors per cluster.
Now we search with nprobe=5, then 5 clusters are scanned, 50000 vectors are computed.

We can see, the number of vectors to be computed are almost the same.
Recall rate is slightly affected because the vectors are re-clustered.
